### PR TITLE
EX-1850 - Implement Report query and deletion

### PR DESCRIPTION
**Pull request by johnsonw**
Implement Report query and deletion in graphql

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2304)

### CODE REVIEW

**Comment by johnsonw** [comment]
image

image

```
[root@adm report]# touch file1.txt
[root@adm report]# ls
file1.txt
[root@adm report]# ls
[root@adm report]#
```

File: iml-api/src/graphql.rs [hunk 1]
```diff
@@ -813,6 +828,20 @@ impl MutationRoot {
...
     }
+
+    /// Delete a stratagem report
+    #[graphql(arguments(filename(description = "The report filename to delete")))]
+    async fn delete_stratagem_report(
+        _context: &Context,
+        filename: String,
+    ) -> juniper::FieldResult<bool> {
+        let mut path = get_report_path();
+        path.push(filename);
```
Comment (jgrund) [inline iml-api/src/graphql.rs:839]: What happens if filename is `../../*`?

File: iml-api/src/graphql.rs [hunk 2]
```diff
@@ -467,6 +469,19 @@ impl QueryRoot {
...
         Ok(xs)
     }
+    /// List Stratagem reports
+    async fn stratagem_reports(
+        _context: &Context,
+    ) -> juniper::FieldResult<Vec<String>> {
+        let paths = tokio::fs::read_dir(get_report_path()).await?;
+
+        let items: Vec<String> = paths
+            .map_ok(|x| x.file_name().into_string().expect("DirEntry to string"))
```
Comment (jgrund) [inline iml-api/src/graphql.rs:479]: Seems like we should return the error instead of panicking here.

File: iml-api/src/graphql.rs (hunk 1 above)
Comment (johnsonw) [reply iml-api/src/graphql.rs:836]: right, I'm currently thinking about that but wanted to push a first draft. Do you have any initial thoughts?

File: iml-api/src/graphql.rs (hunk 2 above)
Comment (johnsonw) [reply iml-api/src/graphql.rs:476]: I'm trying to think of the best way to do this. So far I came up with:
```
let items: Vec<String> = paths
            .map_ok(|x| {
                x.file_name().into_string().map_or_else(
                    |e| format!("Error getting stratagem report file name: {:?}", e),
                    |x| x,
                )
            })
            .try_collect()
            .await?;
```
However, I feel like I'm missing something..

Comment (jgrund) [reply iml-api/src/graphql.rs]: Pushed an update.

File: iml-api/src/graphql.rs (hunk 1 above)
Comment (jgrund) [reply iml-api/src/graphql.rs]: Pushed an update.

**Comment by johnsonw** [comment]
Docker test:

Displaying files:
image

[root@iml-api report]# ls
[root@iml-api report]# ls
expiring_fids-fs-32621d4c-9b38-4d22-8d4a-9a45d96f5667.txt

Deleting a file:
image
```
[root@iml-api report]# ls
expiring_fids-fs-52ee437c-646a-402a-9df3-e90810a77a68.txt
[root@iml-api report]# ls
[root@iml-api report]#
```

**Comment by johnsonw** [comment]
Trying to delete a file maliciously:

image

**Review by mkpankov — APPROVED** [review]
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2304)* status: all files reviewed, 2 unresolved discussions (waiting on @johnsonw)

**Comment by johnsonw** [comment]
Integration tests passed here: https://build.whamcloud.com/view/IML/job/manager-for-lustre/2489/

**Comment by johnsonw** [comment]
Also, I've updated the query to return both the filesize and modified time. The data returned now looks like this:
```
{
  "data": {
    "stratagemReports": [
      {
        "filename": "/var/spool/iml/report/test.txt",
        "modifyTime": "2020-10-05T21:51:57.887528595+00:00",
        "size": 0
      }
    ]
  }
}
```

**Comment by johnsonw** [comment]
The CodeQL failure is unrelated and I see it in other patches as well.

**Comment by jgrund** [comment]
Does this now print the whole file path or just the name?

**Comment by johnsonw** [comment]
It's printing the whole file path. I thought it might be useful to show the whole path in case they weren't aware of the report being under /var/spool/iml. However, we could always print a note above indicating where the file live and then just show the filename. What are your thoughts?

**Comment by jgrund** [comment]
Probably just the filename. The reports can be downloaded at /report/<filename>, so it's probably not important to know the full path. In addition, if we are running in containers, that path will exist inside a container only and will be confusing to users that it's not on the host.

**Comment by johnsonw** [comment]
Sure, makes sense. I will update it to be the filename only.

**Comment by johnsonw** [comment]
Here is an updated screenshot. I created three files under /var/spool/iml/report. Notice that only the filename is now displayed:
test1.txt = 1Kib
test2.txt = 1Mib
test3.txt = 1Gib

image

File: iml-api/src/graphql.rs [hunk 3]
```diff
@@ -813,6 +855,24 @@ impl MutationRoot {
...
+    #[graphql(arguments(filename(description = "The report filename to delete")))]
+    async fn delete_stratagem_report(
+        _context: &Context,
+        filename: String,
+    ) -> juniper::FieldResult<bool> {
+        let report_base = get_report_path();
+        let path = tokio::fs::canonicalize(report_base.join(filename)).await?;
+
+        if !path.starts_with(report_base) {
+            return Err(FieldError::new("Invalid path", Value::null()));
```
Comment (ip1981) [inline iml-api/src/graphql.rs:869]: nit:
```suggestion
            return Err(FieldError::new("Invalid filename", Value::null()));
```

**Review by jgrund — APPROVED** [review]

File: iml-api/src/graphql.rs [hunk 4]
```diff
@@ -467,6 +483,32 @@ impl QueryRoot {
...
+    async fn stratagem_reports(_context: &Context) -> juniper::FieldResult<Vec<StratagemReport>> {
+        let paths = tokio::fs::read_dir(get_report_path()).await?;
+
+        let file_paths = paths
+            .map_ok(|x| x.file_name().to_string_lossy().to_string())
+            .try_collect::<Vec<String>>()
+            .await?
+            .into_iter()
+            .map(|filename| {
+                fs::canonicalize(get_report_path().join(filename.clone()))
```
Comment (ip1981) [inline iml-api/src/graphql.rs:498]: Could call `get_report_path()` juts once at the beginning?

File: iml-api/src/graphql.rs [hunk 5]
```diff
@@ -467,6 +483,32 @@ impl QueryRoot {
 
         Ok(xs)
     }
+
+    /// List completed Stratagem reports that currently reside on the manager node.
+    /// Note: All report names must be valid unicode.
+    async fn stratagem_reports(_context: &Context) -> juniper::FieldResult<Vec<StratagemReport>> {
+        let paths = tokio::fs::read_dir(get_report_path()).await?;
```
Comment (ip1981) [inline iml-api/src/graphql.rs:490]: nit:
```suggestion
        let paths = fs::read_dir(get_report_path()).await?;
```

File: iml-api/src/graphql.rs (hunk 4 above)
Comment (ip1981) [inline iml-api/src/graphql.rs:498]: Is clone() needed?

**Pull request merged**